node/storage-cleaner/src/appwrite.js

```diff
@@ -0,0 +1,98 @@
+import { Client, Storage, Query } from 'node-appwrite';
+import { getExpiryDate } from './util.js';
+
+class AppwriteService {
+  defaultQuery = [Query.orderAsc('$createdAt')];
+
+  constructor() {
+    const client = new Client()
+      .setEndpoint(
+        process.env.APPWRITE_ENDPOINT ?? 'https://cloud.appwrite.io/v1'
+      )
+      .setProject(process.env.APPWRITE_FUNCTION_PROJECT_ID)
+      .setKey(process.env.APPWRITE_API_KEY);
+    this.storage = new Storage(client);
+  }
+
+  /**
+   * Clean up all storage buckets by removing files older than a specified retention period.
+   *
+   * This function retrieves a list of all storage buckets, and for each bucket,
+   * it calls the cleanBucket function to remove files that are older than the calculated expiry date
+   * based on the retention period.
+   *
+   * @returns {Promise<void>} A Promise that resolves when all buckets are cleaned.
+   */
+  async cleanAllBuckets() {
+    const buckets = await this.getBuckets();
+    for (const bucket of buckets) {
+      await this.cleanBucket(bucket.$id);
+    }
+  }
+
+  /**
+   * Retrieves the list of buckets.
+   *
+   * @param {Query[]} queries - An array of queries to filter and sort the results. Defaults to this.defaultQuery.
+   * @param {Cursor|null} cursor - A cursor for pagination. Pass null to start from the beginning.
+   * @returns {Promise<Object[]>} An array of buckets.
+   */
+  async getBuckets(queries = this.defaultQuery, cursor = null) {
+    const currentQueries = [...queries, Query.limit(100)];
+
+    if (cursor) {
+      currentQueries.push(Query.cursorAfter(cursor.$id));
+    }
+    const response = await this.storage.listBuckets(currentQueries);
+    if (response.buckets.length < 100) {
+      return response.buckets;
+    }
+
+    const nextCursor = response.buckets[response.buckets.length - 1];
+    return [
+      ...response.buckets,
+      ...(await this.getBuckets(queries, nextCursor)),
+    ];
+  }
+
+  /**
+   * Clean up files from the storage bucket by removing files older than a specified retention period.
+   *
+   * This function retrieves files from the specified bucket, ordered by creation date in ascending order,
+   * and deletes files that are older than the calculated expiry date based on the retention period.
+   *
+   * @param {string} bucketId - The ID of the storage bucket to clean.
+   * @returns {Promise<void>} A Promise that resolves when the bucket is cleaned.
+   */
+  async cleanBucket(bucketId) {
+    const queries = [Query.orderAsc('$createdAt'), Query.limit(100)];
+    const batchPromises = []; // will use batchPromises to send multiple delete request simultaneously
+    outerLoop: while (true) {
```

Comment:
`while true` is a bit scary. Let's do something along lines:

```js
hasNextPage = true;

do {
  const response = getFIles();
  for(file in response.files) {
   deleteFile(file.$id);
  }
  
  hasNextPage = response.files.length > 0;
} while(hasNextPage);
```

Reply:
here is batching example:

```js
await Promise.all(response.files.map(file) => deleteFile(file.$id));
```

Reply:
Done, the do while change is good, it made the code smaller.